core/runtime/snapshot.rs

```diff
@@ -0,0 +1,577 @@
+      );
+    };
+    let Some(data) = data.take() else {
+      panic!("Attempted to read the snapshot data at index {id} twice");
+    };
+    let local = v8::Local::new(scope, data);
+    let local = v8::Local::<T>::try_from(local).unwrap_or_else(|_| {
+      panic!(
+        "Invalid data type at index {id}, expected '{}'",
+        std::any::type_name::<T>()
+      )
+    });
+    v8::Global::new(scope, local)
+  }
+}
+
+#[derive(Default)]
+pub struct SnapshotStoreDataStore {
+  data: Vec<v8::Global<v8::Data>>,
+}
+
+impl SnapshotStoreDataStore {
+  pub fn register<T>(&mut self, global: v8::Global<T>) -> SnapshotDataId
+  where
+    for<'s> v8::Local<'s, v8::Data>: From<v8::Local<'s, T>>,
+  {
+    let id = self.data.len();
+    // TODO(mmastrac): v8::Global needs From/Into
+    // SAFETY: Because we've tested that Local<Data>: From<Local<T>>, we can assume this is safe.
+    unsafe {
+      self.data.push(std::mem::transmute(global));
+    }
+    id as _
+  }
+}
+
+/// Handles the serialization of the snapshot.
+pub trait SnapshotSerializer: Debug {
+  type Output;
+  fn initialize(&mut self, approximate_length: usize) -> std::io::Result<()>;
+  fn process_chunk(&mut self, chunk: &[u8]) -> std::io::Result<()>;
+  fn finalize(self) -> std::io::Result<Self::Output>;
+}
+
+/// Allows us to consume `self` in a boxed trait. This is required to allow
+/// [`SnapshotSerializer`] to be boxed.
+pub trait SnapshotSerializerSized: SnapshotSerializer {
+  fn finalize_mut(this: &mut Option<Self>) -> std::io::Result<Self::Output>
+  where
+    Self: Sized;
+}
+
+impl<T, S: SnapshotSerializer<Output = T> + Sized> SnapshotSerializerSized
+  for S
+{
+  fn finalize_mut(this: &mut Option<Self>) -> std::io::Result<Self::Output>
+  where
+    Self: Sized,
+  {
+    this.take().unwrap().finalize()
+  }
+}
+
+impl<T> SnapshotSerializer for Box<dyn SnapshotSerializerSized<Output = T>> {
+  type Output = T;
+  fn finalize(self) -> std::io::Result<Self::Output> {
+    Self::finalize_mut(&mut Some(self))
+  }
+  fn initialize(&mut self, approximate_length: usize) -> std::io::Result<()> {
+    (**self).initialize(approximate_length)
+  }
+  fn process_chunk(&mut self, chunk: &[u8]) -> std::io::Result<()> {
+    (**self).process_chunk(chunk)
+  }
+}
+
+/// Serializes the snapshot to a file.
+#[derive(Debug)]
+pub struct SnapshotFileSerializer {
+  file: std::fs::File,
+}
+
+impl SnapshotFileSerializer {
+  pub fn new(file: std::fs::File) -> Self {
+    Self { file }
+  }
+}
+
+impl SnapshotSerializer for SnapshotFileSerializer {
+  type Output = std::fs::File;
+  fn initialize(&mut self, _approximate_length: usize) -> std::io::Result<()> {
+    Ok(())
+  }
+
+  fn process_chunk(&mut self, chunk: &[u8]) -> std::io::Result<()> {
+    self.file.write_all(chunk)
+  }
+
+  fn finalize(self) -> std::io::Result<Self::Output> {
+    Ok(self.file)
+  }
+}
+
+/// Serializes the snapshot to another serializer, with a bulk compression operation performed on top.
```

Comment:
What is bulk compression in this context?

Reply:
I stole the term from `zstd` which we use in the LSP snapshot: `&zstd::bulk::compress(snapshot_slice, 22)`